xla/service/algebraic_simplifier.cc

```diff
@@ -4631,6 +4631,40 @@ Status AlgebraicSimplifierVisitor::HandleCompare(HloInstruction* compare) {
   HloInstruction* rhs;
   CHECK(Match(compare, m::Compare(m::Op(&lhs), m::Op(&rhs))));
 
+  // Gt(Max(a,b), a) -> Gt(b,a)
+  // Gt(Max(a,b), b) -> Gt(a,b)
+  // Gt(a, Min(a,b)) -> Gt(a,b)
+  // Gt(b, Min(a,b)) -> Gt(b,a)
+  if (compare->comparison_direction() == ComparisonDirection::kGt) {
+    HloInstruction* a;
+    HloInstruction* b;
+    if (Match(lhs, m::Maximum(m::Op(&a), m::Op(&b)))) {
+      if (rhs == a) {  // Gt(Max(a,b), a) -> Gt(b,a)
+        TF_ASSIGN_OR_RETURN(auto new_compare,
+                            MakeCompareHlo(ComparisonDirection::kGt, b, a,
+                                           &compare->metadata()));
```

Comment:
Sorry, I just realized that ReplaceInstruction will already copy the metadata over. We would only need to set the metadata for other ops that are not used as direct replacement (not the case here).

Reply:
What is the recommended approach in `Gt(Max(a,b), a) -> Gt(b,a)` case? 
- Use `MakeCompareHlo ` + `ReplaceInstruction(compare, new_compare)` or
- Use `compare->ReplaceOperandWith` + `MarkAsChanged` ?
```
    TF_RETURN_IF_ERROR(compare->ReplaceOperandWith(0, lhs->mutable_operand(1)));
    MarkAsChanged();
```

Reply:
ok, we ended up with using `compare->ReplaceOperandWith` + `MarkAsChanged` for this case. According to Adrian - _It seems ReplaceOperandWith is the better approach that avoids some internal problem with not preserving tiling layouts_